main.go

```diff
@@ -50,23 +50,24 @@ import (
-	webhookSvcName      = flag.String("webhook-svc-name", "spark-webhook", "The name of the Service for the webhook server.")
-	webhookPort         = flag.Int("webhook-port", 8080, "Service port of the webhook server.")
-	enableMetrics       = flag.Bool("enable-metrics", false, "Whether to enable the metrics endpoint.")
-	metricsPort         = flag.String("metrics-port", "10254", "Port for the metrics endpoint.")
-	metricsEndpoint     = flag.String("metrics-endpoint", "/metrics", "Metrics endpoint.")
-	metricsPrefix       = flag.String("metrics-prefix", "", "Prefix for the metrics.")
-	ingressUrlFormat    = flag.String("ingress-url-format", "", "Ingress URL format.")
+	master                       = flag.String("master", "", "The address of the Kubernetes API server. Overrides any value in kubeconfig. Only required if out-of-cluster.")
+	kubeConfig                   = flag.String("kubeConfig", "", "Path to a kube config. Only required if out-of-cluster.")
+	installCRDs                  = flag.Bool("install-crds", true, "Whether to install CRDs")
+	controllerThreads            = flag.Int("controller-threads", 10, "Number of worker threads used by the SparkApplication controller.")
+	resyncInterval               = flag.Int("resync-interval", 30, "Informer resync interval in seconds.")
+	namespace                    = flag.String("namespace", apiv1.NamespaceAll, "The Kubernetes namespace to manage. Will manage custom resource objects of the managed CRD types for the whole cluster if unset.")
+	enableWebhook                = flag.Bool("enable-webhook", false, "Whether to enable the mutating admission webhook for admitting and patching Spark pods.")
+	webhookConfigName            = flag.String("webhook-config-name", "spark-webhook-config", "The name of the MutatingWebhookConfiguration object to create.")
+	webhookCertDir               = flag.String("webhook-cert-dir", "/etc/webhook-certs", "The directory where x509 certificate and key files are stored.")
+	webhookSvcNamespace          = flag.String("webhook-svc-namespace", "spark-operator", "The namespace of the Service for the webhook server.")
+	webhookSvcName               = flag.String("webhook-svc-name", "spark-webhook", "The name of the Service for the webhook server.")
+	webhookPort                  = flag.Int("webhook-port", 8080, "Service port of the webhook server.")
+	enableMetrics                = flag.Bool("enable-metrics", false, "Whether to enable the metrics endpoint.")
+	metricsPort                  = flag.String("metrics-port", "10254", "Port for the metrics endpoint.")
+	metricsEndpoint              = flag.String("metrics-endpoint", "/metrics", "Metrics endpoint.")
+	metricsPrefix                = flag.String("metrics-prefix", "", "Prefix for the metrics.")
+	ingressUrlFormat             = flag.String("ingress-url-format", "", "Ingress URL format.")
+	sparkUIServiceAnnotationsStr = flag.String("spark-ui-service-annotations", "{}", "Annotations (as JSON string) for sparkui service.")
```

Comment:
Is it possible that different applications will need different annotations, even in the same cluster?

Reply:
Yes and no. For the spark job use case, I can only think of 3 types of annotations I might need:

- pod annotations (driver/executor pods) for kube2iam, kiam, etc (credentials for access control) (already provided in the CRD)

- service annotations (sparkUI) for ambassador which need it to create the envoy proxy (this PR)

- ingress annotations (sparkUI) for other custom ingress controllers (e.g. nginx, kong, alb) also relies on these annotations to create the proxy/entry point. However, u probably use either ambassador or a custom ingress controller (they both serve the same purpose)

PS: This probably means I should also add annotation for the ingress to support pple who use custom ingress controllers.

Reply:
Just reread ur question and realised that u are asking whether annotation should be specified at the CRD (aka diff spark jobs need diff annotations).

I don't think the annotations need to change if it is templated (diff route for diff spark job). However, if we do it at the CRD, then we probably don't need templating. 

But yeah, now that u mentioned it. It might be easier to write it at the CRD. json string as arg is abit clunky. And might not be obvious to the end user, as they might not be the one who setup the k8s spark operator. 

But on the other hand, I might need to teach my data scientists how ambassador works, which may not be desirable, and probably also make the yaml less pure (the annotation is purely a cluster specific setup).

Reply:
Makes sense.